From ba3a32b3b6907a36b55b6fb56477cde53e6d46f0 Mon Sep 17 00:00:00 2001
From: Richard Zhu <hongxing.zhu@nxp.com>
Date: Thu, 4 May 2017 14:58:35 +0800
Subject: [PATCH 1762/5242] MLK-14796 PCI: imx: kernel panic when reboot ep
 board

commit  82ebd6f8d44e988d94c0287c4b5620a9aa23575e from
https://source.codeaurora.org/external/imx/linux-imx.git

pp->ops is not assigned properly, thus there
would be kernel panic when reboot ep board.
Initialized pp->ops in ep initializatione,
fix this issue.
don't call dw_pcie_wait_for_link because
the usleep is used in it.

Signed-off-by: Richard Zhu <hongxing.zhu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/pci/controller/dwc/pci-imx6.c |   21 ++++++++++++++-------
 1 file changed, 14 insertions(+), 7 deletions(-)

diff --git a/drivers/pci/controller/dwc/pci-imx6.c b/drivers/pci/controller/dwc/pci-imx6.c
index 2731966..dd8f869 100644
--- a/drivers/pci/controller/dwc/pci-imx6.c
+++ b/drivers/pci/controller/dwc/pci-imx6.c
@@ -664,18 +664,24 @@ static void imx6_pcie_init_phy(struct imx6_pcie *imx6_pcie)
 
 static int imx6_pcie_wait_for_link(struct imx6_pcie *imx6_pcie)
 {
+	int count = 20000;
 	struct dw_pcie *pci = imx6_pcie->pci;
 	struct device *dev = pci->dev;
 
 	/* check if the link is up or not */
-	if (!dw_pcie_wait_for_link(pci))
-		return 0;
-
-	dev_dbg(dev, "DEBUG_R0: 0x%08x, DEBUG_R1: 0x%08x\n",
-		dw_pcie_readl_dbi(pci, PCIE_PHY_DEBUG_R0),
-		dw_pcie_readl_dbi(pci, PCIE_PHY_DEBUG_R1));
+	while (!dw_pcie_link_up(pci)) {
+		udelay(10);
+		if (--count)
+			continue;
+
+		dev_err(dev, "phy link never came up\n");
+		dev_dbg(dev, "DEBUG_R0: 0x%08x, DEBUG_R1: 0x%08x\n",
+			dw_pcie_readl_dbi(pci, PCIE_PHY_DEBUG_R0),
+			dw_pcie_readl_dbi(pci, PCIE_PHY_DEBUG_R1));
+		return -ETIMEDOUT;
+	}
 
-	return -ETIMEDOUT;
+	return 0;
 }
 
 static int imx6_pcie_wait_for_speed_change(struct imx6_pcie *imx6_pcie)
@@ -1445,6 +1451,7 @@ static int __init imx6_pcie_probe(struct platform_device *pdev)
 		}
 
 		pp->mem_base = pp->mem->start;
+		pp->ops = &imx6_pcie_host_ops;
 
 		/* enable disp_mix power domain */
 		if (imx6_pcie->variant == IMX7D)
-- 
1.7.9.5

